src/main/c/Region.cpp

```diff
@@ -0,0 +1,111 @@
+#include "hail/Region.h"
+#include "hail/NativeObj.h"
+#include "hail/NativePtr.h"
+#include <jni.h>
+
+namespace hail {
+
+void Region::clear_but_keep_mem() {
+  pos_ = (chunk_.empty() ? kChunkCap : 0);
+  for (auto& full : full_chunks_) free_chunks_.emplace_back(std::move(full));
```

Comment:
So my original goal in the design of region was (1) make the common allocation path very fast (comparison plus pointer bump) and make the deallocation O(1).  I agree moving around the blocks isn't ideal looking code, but I don't think we want to do it either: blocks are always allocated in linear order, so if we keep a list of chunks and a single index which is the current chunk, `clear_but_keep_mem` for the chunks just means setting that index and the pos to 0.  Then all the moves related to chunks go away.

Reply:
Yes, but that then involves an extra level of indirection each time we access the current
chunk buf - which is on the critical path of fast-allocate.  I guess we could avoid that by also
keeping the char* cur_chunk_buf as its own thing, but then we've complicated the
new_chunk code again.

Reply:
I think the latest code meets your goals, but by keeping the current buf_ at the top level it doesn't do any unnecessary indirection.